confidant/services/jwkmanager.py

```diff
@@ -8,29 +8,90 @@
 from typing import Tuple
 
 import jwt
+from abc import ABC, abstractmethod
 from cerberus import Validator
 from confidant.settings import JWT_ACTIVE_SIGNING_KEYS
 from confidant.settings import JWT_CACHING_ENABLED
 from confidant.settings import JWT_CERTIFICATE_AUTHORITIES
 from confidant.settings import JWT_DEFAULT_JWT_EXPIRATION_SECONDS
+from confidant.settings import JWT_CACHING_MAX_SIZE
+from confidant.settings import JWT_CACHING_TTL_SECONDS
 from confidant.utils import stats
-from jwcrypto import jwk
 
+from cachetools import TTLCache
+from jwcrypto import jwk
 
 logger = logging.getLogger(__name__)
 
 CA_SCHEMA = {
     'crt': {'type': 'string', 'required': True},
     'key': {'type': 'string', 'required': True},
-    'passphrase': {'type': 'string', 'required': True},
+    'passphrase': {'type': 'string', 'required': True, 'nullable': True},
     'kid': {'type': 'string', 'required': True},
 }
 
 
+class JwtCache(ABC):
+    def statsdcache(f):
+        """ Decorator to send stats based on result of cache hits/misses
+        """
+        def wrapper(self, *args, **kwargs):
+            result = f(self, *args, **kwargs)
+            if result:
+                stats.incr(f'jwt.get_jwt.cache.{kwargs["kid"]}.hit')
+            else:
+                stats.incr(f'jwt.get_jwt.cache.{kwargs["kid"]}.miss')
+            return result
+        return wrapper
+
+    @abstractmethod
+    def get_jwt(self, kid: str, requester: str, user: str) -> str:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def set_jwt(self, kid: str, requester: str, user: str, jwt: str) -> None:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def cache_key(self, kid: str, requester: str, user: str) -> str:
```

Comment:
nit: Probably not needed as in abstract base class as it can be an internal function of derived classes

Reply:
removed